utils/filecache/cache.go

```diff
@@ -0,0 +1,200 @@
+package filecache
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	fs "github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+)
+
+type Cache struct {
+	entries       *entryMap
+	entriesLM     *lockMap
+	entryProvider IEntryProvider
+	fs            fs.FS
+	cfg           *FileCacheConfig
+	cancelStore   *parallelisation.CancelFunctionStore
+	closed        atomic.Bool
+}
+
+func (c *Cache) gc(ctx context.Context, _ time.Time) {
+	c.entries.Range(func(key string, entry ICacheEntry) {
+		if entry.IsExpired() && c.entriesLM.TryLock(key) {
+			if err := entry.Delete(ctx, c.fs); err == nil {
+				c.entries.Delete(key)
+				c.entriesLM.Unlock(key)
```

Comment:
It is good to use defer for unlocks etc. It might be fine now, but in the event we add more logic it could cause issues

```suggestion
			defer c.entriesLM.Unlock(key)
			if err := entry.Delete(ctx, c.fs); err == nil {
				c.entries.Delete(key)
```